hotshot-state-prover/src/service.rs

```diff
@@ -0,0 +1,575 @@
+        .expect("unable to instantiate Provider, likely wrong URL");
+    let signer = Wallet::from(config.eth_signing_key.clone())
+        .with_chain_id(provider.get_chainid().await?.as_u64());
+    let l1_wallet = Arc::new(L1Wallet::new(provider, signer));
+
+    let contract = LightClient::new(config.light_client_address, l1_wallet);
+    Ok(contract)
+}
+
+/// get the `finalizedState` from the LightClient contract storage on L1
+pub async fn read_contract_state(
+    config: &StateProverConfig,
+) -> Result<LightClientState, ProverError> {
+    let contract = prepare_contract(config).await?;
+    let state: ParsedLightClientState = match contract.finalized_state().call().await {
+        Ok(s) => s.into(),
+        Err(e) => {
+            tracing::error!("unable to read finalized_state from contract: {}", e);
+            return Err(ProverError::ContractError(e.into()));
+        }
+    };
+    let state: LightClientState = state.into();
+
+    Ok(state)
+}
+
+/// submit the latest finalized state along with a proof to the L1 LightClient contract
+pub async fn submit_state_and_proof(
+    proof: Proof,
+    public_input: PublicInput,
+    config: &StateProverConfig,
+) -> Result<(), ProverError> {
+    let contract = prepare_contract(config).await?;
+
+    // prepare the input the contract call and the tx itself
+    let proof: ParsedPlonkProof = proof.into();
+    let new_state: ParsedLightClientState = public_input.into();
+    let tx = contract.new_finalized_state(new_state.into(), proof.into());
+
+    // send the tx
+    let (_receipt, _included_block) = sequencer_utils::contract_send(&tx)
+        .await
+        .map_err(ProverError::ContractError)?;
+
+    Ok(())
+}
+
+pub async fn sync_state(
+    st: &StakeTable<BLSPubKey, StateVerKey, CircuitField>,
+    proving_key: &ProvingKey,
+    relay_server_client: &Client<ServerError>,
+    config: &StateProverConfig,
+) -> Result<(), ProverError> {
+    tracing::info!("Start syncing light client state.");
+
+    let bundle = fetch_latest_state(relay_server_client).await?;
+    let old_state = read_contract_state(config).await?;
+    if old_state.block_height >= bundle.state.block_height {
+        tracing::info!("No update needed.");
+        return Ok(());
+    }
+
+    let threshold = st.total_stake(SnapshotVersion::LastEpochStart)? * 2 / 3;
+    let entries = st
+        .try_iter(SnapshotVersion::LastEpochStart)
+        .unwrap()
+        .map(|(_, stake_amount, state_key)| (state_key, stake_amount))
+        .collect::<Vec<_>>();
+    let mut signer_bit_vec = vec![false; entries.len()];
+    let mut signatures = vec![Default::default(); entries.len()];
+    let mut accumulated_weight = U256::zero();
+    entries.iter().enumerate().for_each(|(i, (key, stake))| {
+        if let Some(sig) = bundle.signatures.get(key) {
+            signer_bit_vec[i] = true;
+            signatures[i] = sig.clone();
+            accumulated_weight += *stake;
+        }
+    });
+
```

Comment:
Maybe throw an error if `accumulated_weight < threshold` instead of getting the error while generating the proof.

Reply:
Done in [f005900](https://github.com/EspressoSystems/espresso-sequencer/pull/1041/commits/f0059003d42ca79309513fe94c1ffc647d3e474b)